### PR TITLE
Add defaults

### LINKED ISSUE #48: Also annotate parameter defaults.
In numpy docstrings I often see

```rst
Parameters
----------
myparam : `Type`, optional (default: `None`)
    My param does things.
```

I prefer styling the parameter docs as definition list like numpydoc, where there is enough space for type and defaults:

<blockquote>
<dl>
<dt><strong>myparam</strong>: <code>Type</code>, optional (default: <code>None</code>)</dt>
<dd>My param does things.</dd>
</dl>
</blockquote>

However, the defaults get lost when migrating to sphinx-autodoc-typehints. I know the defaults are in the function signature above the param docs, but for long signatures (and therefor long lists of params), it would be best to also have them in the param docs.

Ideally we would parse and use the source code for the defaults instead of going from python objects back to strings, but using `repr` to get OK representations for most defaults would also be OK.

### PR DESCRIPTION
**Pull request by flying-sheep**
Supersedes #103, fixes #48

In the previous PR, I added them to types, this time I’m adding it to the string. I think both have arguments pro and contra. Should it be configurable?

### CODE REVIEW

**Comment by gaborbernat** [comment]
I'd like to have it by default off. Otherwise, I'm happy to also allow it to be configurable where it's added.

**Comment by flying-sheep** [comment]
OK, I need help, I have no idea what’s going on with those `"`s and `*`s.

**Comment by gaborbernat** [comment]
> OK, I need help, I have no idea what’s going on with those `"`s and `*`s.
> 
> Why does only my test case fail on Python 3.9 even though I copied the setup from `test_sphinx_output_future_annotations`?

That's half the work in this project, to figure out the differences between Python versions and make it work everywhere. I'll have a look at some point, but no idea when I'll have time for this.

**Comment by flying-sheep** [comment]
Ugh, I figured it out. This will bite you in the future! So even on 3.9, a simple type like `int` gets compiled to

- `*int*` if the file *contains* new-style union syntax, and
- `"int"` otherwise.

It’s *not* about `from __future__ import annotations`, it’s about actually using the syntax! So:

```py
from __future__ import annotations

def function(x: bool) -> str | int: ...
```

gets compiled to `*`

```
...
Parameters:
   * **x** (*bool*) -- foo
```

And when you don‘t use `|` syntax:

```py
from __future__ import annotations

def function(x: bool) -> str: ...
```

gets compiled to `"`

```
...
Parameters:
   * **x** ("bool") -- foo
```

**Comment by gaborbernat** [comment]
> Ugh, I figured it out. This will bite you in the future, I hope you find a way to make it less flaky.
> 
> I filed #194 to explore this. But this PR is now ready!

Sorry, I'm not open to a fallback issue and working around the test. We should figure out here and now why this is the case and avoid it.

**Comment by flying-sheep** [comment]
I’m avoiding it by not using that syntax, in the same way your main test does. This PR is completely independent of this problem now, I just happened to run into it while developing this PR.

File: src/sphinx_autodoc_typehints/__init__.py [hunk 1]
```diff
@@ -493,16 +516,20 @@ def process_docstring(app, what, name, obj, options, lines):  # noqa: U100
                 lines.insert(insert_index, f":rtype: {formatted_annotation}")
 
 
-def builder_ready(app):
+def builder_ready(app: Sphinx):
```
Comment (gaborbernat) [inline src/sphinx_autodoc_typehints/__init__.py:519]: Let's type-hint the return value to if you're going to type annotate it.

File: src/sphinx_autodoc_typehints/__init__.py [hunk 2]
```diff
@@ -493,16 +516,20 @@ def process_docstring(app, what, name, obj, options, lines):  # noqa: U100
...
     if app.config.set_type_checking_flag:
         typing.TYPE_CHECKING = True
+    valid = {None, "comma", "braces", "braces-after"}
+    if app.config.typehints_defaults not in valid | {False}:
+        raise ValueError(f"typehints_defaults needs to be one of {valid!r}, not {app.config.typehints_defaults!r}")
```
Comment (gaborbernat) [inline src/sphinx_autodoc_typehints/__init__.py:524]: test for this please

File: src/sphinx_autodoc_typehints/__init__.py [hunk 3]
```diff
@@ -425,7 +426,19 @@ def add(val):
     return result
 
 
-def process_docstring(app, what, name, obj, options, lines):  # noqa: U100
+def format_default(app: Sphinx, default):
+    if not app.config.typehints_defaults:
+        return None
```
Comment (gaborbernat) [inline src/sphinx_autodoc_typehints/__init__.py:431]: If this is the case the method will never be called so I don't think we need this

**Review by gaborbernat — CHANGES_REQUESTED** [review]

**Pull request merged**